sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_marray.asciidoc

```diff
@@ -0,0 +1,379 @@
+link:sycl_ext_oneapi_complex.asciidoc[sycl_ext_oneapi_complex].
+
+
+== Status
+
+This is a proposed extension specification, intended to gather community
+feedback.  Interfaces defined in this specification may not be implemented yet
+or may be in a preliminary state.  The specification itself may also change in
+incompatible ways before it is finalized.  *Shipping software products should
+not rely on APIs defined in this specification.*
+
+[NOTE]
+====
+This extension is not currently implemented in {dpcpp}.
+====
+
+
+== Overview
+
+{dpcpp} has support for `sycl::ext::oneapi::complex` this allows for the
+addition of new complex math features. This proposes the instantiation of
+`marray` to add support for complex numbers to be stored in arrays. The
+proposal includes overloading the existing math functions to support complex
+marrays and adding new free functions to simplify accessing, setting marray
+values, and construct complex marrays.
+
+== Specification
+
+=== Feature test macro
+
+This extension provides a feature-test macro as described in the core SYCL
+specification.  An implementation supporting this extension must predefine the
+macro `SYCL_EXT_ONEAPI_COMPLEX_MARRAY` to one of the values defined in the table
+below.  Applications can test for the existence of this macro to determine if
+the implementation supports this feature, or applications can test the macro's
+value to determine which of the extension's features the implementation
+supports.
+
+[%header,cols="1,5"]
+|===
+|Value
+|Description
+
+|1
+|Initial version of this extension.
+|===
+
+=== Marray Complex Class Specialization
+
+The `marray` class is extended for the `sycl::ext::oneapi::complex` class.
+The user interface of the `marray<sycl::ext::oneapi::complex, {N}>`
+class is not changed compared to the SYCL-2020 generic `marray` interface.
+
+The `marray` complex instantiation is trivially copyable, and the type
+trait `is_device_copyable` should resolve to `std::true_type`.
+
+As the `marray` class is not specialized in this proposal. The `marray`
+definition used within this proposal assumes that any operator the `marray`
+class defines is only implemented if the marray's value type also
+implements the operator. So, for example, `marray<float, N>` does not
+implement the modulus operator as float does not support it. As the
+`marray` class is not changed in this proposal, it is not redefined below.
+
+The `make_complex_marray` free function is added to construct complex
+marrays from real and imaginary components. Additionally, the free
+functions `get_real` and `get_imag` are added to access the real and
+imaginary components of the `marray` class without modifying the existing
+`marray` interface. The usage of free functions does cause a deviation
+from the `std::complex` interface. However, it does reduce this extensions
+impact on the `marray` interface.
+
+```C++
+namespace sycl {
+namespace ext {
+namespace oneapi {
+
+  // Make_complex_marray
```

Comment:
Should we really add multiple `make_complex_marray` overloads just for complex? Wouldn't it make more sense to add a general purpose generator expression mechanism? E.g. like the one of std::simd in parallelism TSv2.
If we add `template<class G> marray(G&& gen)` instead of writing `make_complex_marray(ar, ai)` (where ar, ai are marrays) one could write `marray([&](auto i){return complex{ar[i], ai[i]};}`. Isn't that simple enough?

Reply:
Applications using arrays of complex values will likely want an interface that performs similar operations to the `std::complex` interface. I.e. being able to construct a complex value with a real and imaginary vector component. While `marray([&](auto i){return complex{ar[i], ai[i]};}` would work, it is much more verbose than `make_complex_marray(ar, ai)`, for something that is a simple element-wise constructor and it would be a significant departure from the simple `std::complex` constructors.

Reply:
> Applications using arrays of complex values will likely want an interface that performs similar operations to the std::complex interface.

I can see that. The question is why should that need be met by SYCL (initially as extension)?I don't think we want to add all math functionality applications reasonable want to have and slowly end up with an API size similar to libraries like Eigen. To keep the API reasonable small we should only have in SYCL what can be considered fundamental building blocks and/or can't be implemented (efficiently) as library.  And I don't think `make_complex_marray` qualifies as such.

Reply:
So regarding your point, I think that if these functions were made member functions it may be less intrusive and awkward in the spec. I have updated the proposal to specialize `marray` and made these functions members of the class. Let me know your thoughts on this approach and if this addresses some of your concerns.